proxy/http2/Http2ConnectionState.cc

```diff
@@ -1790,19 +2090,20 @@ Http2ConnectionState::send_a_data_frame(Http2Stream *stream, size_t &payload_len
   }
 
   if (stream->is_write_vio_done()) {
+    Http2StreamDebug(this->session, stream->get_id(), "End of Data Frame");
     flags |= HTTP2_FLAGS_DATA_END_STREAM;
   }
 
   // Update window size
-  this->decrement_peer_rwnd_in(payload_length);
+  this->decrement_peer_rwnd(payload_length);
   stream->decrement_peer_rwnd(payload_length);
 
   // Create frame
-  Http2StreamDebug(session, stream->get_id(), "Send a DATA frame - client window con: %5zd stream: %5zd payload: %5zd",
-                   _peer_rwnd_in, stream->get_peer_rwnd(), payload_length);
+  Http2StreamDebug(session, stream->get_id(), "Send a DATA frame - peer window con: %5zd stream: %5zd payload: %5zd flags: 0x%x",
+                   _peer_rwnd, stream->get_peer_rwnd(), payload_length, flags);
 
   Http2DataFrame data(stream->get_id(), flags, resp_reader, payload_length);
-  this->session->xmit(data, flags & HTTP2_FLAGS_DATA_END_STREAM);
+  this->session->xmit(data);
```

Comment:
I don't understand why we can remove the second parameter for END_STREAM flag.

Reply:
The second parameter is whether to flush, which defaults to true. In the previous patch we only flushed when the END_STREAM flag was set. I found situations via http2_flow_control.test.py where this would cause hanging if we didn't flush with each of these calls. So I changed it to always flush.

Reply:
I think I added it to reduce the number of SSL_write calls with small payloads, and the buffer should be flushed if a certain amount of time elapsed since the last flush. And it's on 9.2.x. If there's a possibility of stall we should fix it separately so we can back port it to 9.2.x.

Reply:
I didn't see issues until I added the extra http2.flow_control mechanisms that came about with this PR. Once things became more complicated between handling DATA frames and flow control states on both the client and server side, then I noticed that not always flushing was a problem. I don't think 9.2.x currently needs this, therefore.

Reply:
How delaying flush causes hanging? Sounds like the time-based flush logic is broken. Perhaps some event that used to be scheduled is no longer scheduled with this change and that takes away the opportunity to flush. I'm more worried about the behavior change than performance.

Reply:
After some investigation, I was able to restore the conditional flush by removing the HTTP/2 WINDOW check logic that was added via this patch to `Http2Stream::update_write_request`. This passes all h2 flow control tests.

@shinrich : let me know whether you have any concerns about the `Remove window check from update_write_request` commit I just added for this PR. Currently that is this commit:
https://github.com/apache/trafficserver/pull/9366/commits/7c65be82299d491da076c452f6601eee3ab9e324